src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/LambdaCompiler.Unary.cs

```diff
@@ -333,22 +333,22 @@ private void EmitConvert(UnaryExpression node, CompilationFlags flags)
                         paramType = paramType.GetElementType();
                     }
 
-                    UnaryExpression e = Expression.Convert(
+                    node = Expression.Convert(
                         Expression.Call(
                             node.Method,
                             Expression.Convert(node.Operand, paramType)
                         ),
                         node.Type
                     );
-
-                    EmitConvert(e, flags);
```

Comment:
The conversion expressions created here won't have a method, so the branch at 346 immediately below (or indeed, re-doing this branch another time) won't be a possibility. Changing the `else if` to `if` at line 351 means it will reach the necessary path, and adding the `return` at 347 in turn means that `if` won't now be hit when it shouldn't be.